### ISSUE TITLE
[CoE Starter Kit - QUESTION] Microsoft Teams Admin Weekly Cleanup Issue

### ISSUE THREAD

**Issue by JayCecchine** [microsoft/coe-starter-kit#4036]
### What is your question?

I am implementing governance component in our power platform tenant, to give some context we already had ~330 environments or so.  They were created ad hoc in the past two years.  I ran the the Microsoft Teams Admin | Weekly Clean Up of Microsoft Teams environments flow and it deleted 146 environments.  I believe it deleted them because they had no business justification, as that was not part of the process since CoE was never fully implemented in this tenant.

I spent the better part of the day recovering environments and have turned this flow off for now.  I need some guidance on how to find out why they were deleted and also how to force the existing environments to have a business justification.  I see there is a flow called "Microsoft Teams Admin | Ask for Business Justification when Microsoft Teams environment is created"

But that is just when they're created... I need to enforce on pre-existing environments.  Also it would be nice to be able to run the teams admin cleanup flow in audit mode to see *What* would happen rather than it executing.

Thank you for all your help

### What solution version are you using?

October 2022

### What app or flow are you having the issue with?

Microsoft Teams Admin | Weekly Clean Up of Microsoft Teams environments

**Comment by Jenefer-Monroe** [comment]
Hello. First let's talk through what the expected behavior is. 
The flow does not work on pre-existing environments and we don't actually have an out of the box solution for that for people. Our recommendation is to enable the feature for new teams environments and then to do a one-off clean up of pre-existing ones. 
The only environments we delete for you in clean up are empty teams environments (meaning no apps and flows) which are 90 days or older. 

Now as to why pre-existing environments were deleted, I'm really not sure. They should not be. 
Can you please open the **_Microsoft Teams Admin  | Weekly Clean Up of Microsoft Teams environments_** flow in edit mode and validate that you see the following for the filter on this action
![image](https://user-images.githubusercontent.com/89584709/197863220-2d630cf6-d175-423b-99bd-3858354cb2d6.png)

In particular, this clause in the filter is what protects these pre-existing environments as that field is not filled except by the business justification flow. 
`admin_businessjustificationdate ne null`

**Comment by JayCecchine** [comment]
Also I know this is a separate issue and if you want me to make a separate ticket I will. 

But the Business justification flow has been running for over 18 hours and is stuck on the following, is this normal? -

![image](https://user-images.githubusercontent.com/24686498/197867609-c3e486e2-c6c8-4bcf-955f-a85ca8f87b2d.png)

**Comment by Jenefer-Monroe** [comment]
Yes thats normal to run a long time. Its posted the card to the envt maker and is waiting for them to provide their justification. It will time out if they dont respond in 7 days and their envt will be deleted by the other flow.

**Comment by Jenefer-Monroe** [comment]
your filters look correct. Perhaps the date asked for justification field got filled in?

**Comment by JayCecchine** [comment]
Where would that date asked for justification field need to be filled in at?

**Comment by JayCecchine** [comment]
Also, it is waiting for the user to respond to the card, can I see which environment / user it is hung up on?  Does it get hung up on one user at a time or is it waiting on multiple users to respond (hopefully)

**Comment by Jenefer-Monroe** [comment]
You should not fill the data manually. Thats why I asked. I was trying to help figure out why the teams envts got deleted. Its filled by the Ask for Business Justification flow

**Comment by Jenefer-Monroe** [comment]
If you look in the trigger it will tell you what envt this is for. It runs once per new D4T environment so one user not answering for their envt will not block another user from getting asked or answering. That will be a separate run.

**Comment by JayCecchine** [comment]
Understood, I am having many issues since implementing the CoE kit on this tenant.  This basic add maker to maker group is not functioning - 

![image](https://user-images.githubusercontent.com/24686498/198115355-0e34a408-d80a-499e-8594-12cb5d7290c5.png)

**Comment by JayCecchine** [comment]
Disregard that last question, I needed to add the GUID to the env variable - f42e03d5-1e68-4f54-9b25-9f9473218f68

It is working now

**Comment by JayCecchine** [comment]
Back to the original issue on this ticket, how do I know its going to be safe to run the weekly teams cleanup flow again?

You verified yourself that nothing is wrong with the flow.

Thanks

**Comment by Jenefer-Monroe** [comment]
I'm not sure. I've not heard of other people having D4T envts accidently deleted, so I don't know how that happened in your case, or if its even related to our flows. Do you have a flow run showing our flows being what deleted them? If so, that would help us debug what happened.

**Comment by JayCecchine** [comment]
Below are the flow runs, I'm not sure how to check exactly what was deleted etc, but I doubt its any coincidence the environments were deleted after this particular flow ran from our CoE environment.  I have no custom flows running that would delete environments.  Just what is included from the CoE Starter Kit.

![image](https://user-images.githubusercontent.com/24686498/198392442-9a39ac4a-a2ec-4343-90b7-af3392e73746.png)

**Comment by JayCecchine** [comment]
I was able to get the results of the "failed" run there - 

![image](https://user-images.githubusercontent.com/24686498/198397263-0faf2c54-4b9a-4fed-ac6d-a86ebd57c8e7.png)

There are 170 entries here, not all of them failed.  some of the deletions went through.  the environment ids are all different.  Some of these environments are two years old

**Comment by JayCecchine** [comment]
Here is some additional info and insight perhaps as to why they were deleted.  I went to my environment request review, and there are 182 environments that are up for review with no business justification.  It doesn't look like they were ever required to have one but they're up for a review.

Since these environments were created to years ago, is there a way to send all 182 owners requests for business justification?  Going through this manually and reaching out will be an absolute nightmare.  Please see below - 

![image](https://user-images.githubusercontent.com/24686498/198403568-fc557480-cdeb-487c-9650-0c6776ff194e.png)

**Comment by Jenefer-Monroe** [comment]
Ok. So it deleted them for being non-active. That's good to know. We only have insight in to D4T flows and apps here, we cant look for tables or chatbots etc. so if you want these environments to stay around even though they are "empty" from the POV of this flow, then you will excuse them via the **_Excuse From Empty Environment Cleanup_** field

You can change that in the Power Platform Admin View app

(how it looks in current release)
![image](https://user-images.githubusercontent.com/89584709/198403716-06f36430-cafb-49c6-bdde-e98f141ecceb.png)

(how it looks as of upcoming Nov release)
![image](https://user-images.githubusercontent.com/89584709/198403476-e0403ccd-ffc4-44d2-9400-c385fd75b0ce.png)

**Comment by Jenefer-Monroe** [comment]
They did not get deleted for their business justification being unfilled, thats only for new envts, we dont have a solution for asking for business justification for old environments. They were deleted because they looked empty to us and were over 90 days old.

**Comment by JayCecchine** [comment]
yes all the environments were microsoft teams environments.  Won't this problem happen again on new environments created, since as you said they could be using tables / chat bots.  It will just delete them after 90 days.

I currently have this flow turned off, I guess I should just keep it that way?

**Comment by Jenefer-Monroe** [comment]
This will happen for D4T envts that do not have apps or flows after 90 days. Its not actually common from my understanding but perhaps your tenant has it more regularly. 
There is a new product feature coming which will replace this unused D4T part of the flow in another month or so and its more sophisticated so you may want to look for that.

**Comment by Jenefer-Monroe** [comment]
closing out as no further action for starter kit team
